EIPS/eip-7873.md

```diff
@@ -155,21 +155,41 @@ At the start of the block in which [EIP-3540](./eip-3540.md) activates, set the
 
 ```solidity
 {
-    /// Takes [tx_initcode_hash][salt][init_data] as input,
-    /// creates contract and returns the address or failure otherwise
+    /// Takes [tx_initcode_hash][salt][unsafe_flag][init_data] as input,
+    /// creates contract and returns the address or failure otherwise.
 
-    // init_data.length can be 0, but the first 2 words are mandatory
-    let size := calldatasize()
-    if lt(size, 64) { revert(0, 0) }
+    /// tx_initcode_hash and salt are 32 bytes wide, unsafe_flag is 1 byte wide, init_data can be any width.
+    /// If unsafe_flag is 0x00, `CALLER` is included in the salt and impacts the target address of the created account.
+    /// If unsafe_flag is non-zero, `CALLER` is excluded. Deployment by any sender will result in same target address. 
+    /// Deployments using non-zero unsafe_flag are susceptible to front-running.
 
-    calldatacopy(0, 0, size)  // copy tx_initcode_hash, salt and init_data to memory to hash
-    let final_salt := keccak256(0, size)
+    // init_data.length can be 0, but the first 65 bytes are mandatory
+    let size := calldatasize()
+    if lt(size, 65) { revert(0, 0) }
+
+    // copy tx_initcode_hash, salt, unsafe_flag and init_data to memory to hash.
+    // note that the unsafe_flag is included to prevent hash collisions between the two branches:
+    //   - unsafe_flag == 0, caller's address is appended by the code
+    //   - unsafe_flag == 1, caller's address is sent via init_data
+    calldatacopy(0, 0, size)
+
+    // mask out the 31 bytes that follow the flag in input data
+    let unsafe_flag := byte(0, calldataload(64))
+
+    if iszero(unsafe_flag) {
+        // Include sender into salt just after init_data
+        mstore(size, caller())
+        let final_salt := keccak256(0, size + 32)
```

Comment:
This is insecure. Even if you include the address of the caller this can be front-runned by:
- setting unsafe_flag to 1,
- appending the caller's address to the input.

I think the better approach is to have a fixed placeholder of 32-bytes for the caller's address before the input data. If `unsafe_flag` is set leave the placeholder all zeros, otherwise write caller's address there.

Reply:
Alternatively, always hash over `size + 32` but only write the sender in the last word if `unsafe_flag` is nonzero.

Reply:
`unsafe_flag` is included in the hashing, so that front-running won't work (will land in different target address), and there's a comment laying out this exact reason (also raised on discord and explained).

> // note that the unsafe_flag is included to prevent hash collisions between the two branches

Am I maybe missing sth?

Reply:
No that sounds right... If a tx is "frontrun" as @chfast suggests, it should result in a different deployed address.

Reply:
I pushed the proposed alternative to 0b55a4f5. Note that this is slightly more complex actually, as long as we keep the earlier suggestion to place `unsafe_flag` as third argument. We can debate whether or not to keep it there.

Overall it has a nice property of cutting out the address malleability via various non-zero values of `unsafe_flag` byte...

Reply:
@gumb0 @chfast LMK if you're happy with the latest revision and :heavy_check_mark: if yes, so we can merge this if there's also consensus on the overall direction.

Reply:
I have more ideas in https://github.com/ipsilon/eof/issues/172#issuecomment-2732366390, but maybe better to make another PR to present my new ideas :)

So I think this is fine to merge.